### PR TITLE
feat: add C API to `@stdlib/math/base/special/negalucas`

### LINKED ISSUE #893: [RFC]: Add C implementation for `@stdlib/math/base/special/negalucas`
### Description

This RFC proposes adding C implementation for [`@stdlib/math/base/special/negalucas`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/negalucas).

```c
int stdlib_math_negalucas( const int x );
```

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Jordan-Gallivan**
Resolves #893 .

## Description

> What is the purpose of this pull request?

This pull request: Completed the C negative lucas implementation

## Related Issues

> Does this pull request have any related issues?

No.  All tests completed.  All files compiled

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by Jordan-Gallivan** [comment]
yikes.  Looks like it failed quite a few of the PR checks.  Please let me know what I can do to remedy the above.

**Comment by kgryte** [comment]
@Jordan-Gallivan Thanks for this PR. Looking at the CI checks, they don't appear to be due to anything you've done. We're still working out some issues with our CI pipelines. I'll do a more extensive review within the next 24-48 hours.

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile:4]:
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile]: Are all these changes needed for this PR? I wonder I have not updated any Makefile while adding C implementations.

Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile]: Add a newline.

Comment (Jordan-Gallivan) [reply lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile]: @Pranavchiku I could not get the benchmark to compile with the makefiles that were in the repository.  I pulled the one that was successful from tribonacci and was able to get all files to compile, and all tests to pass

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/native/benchmark.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/native/benchmark.c]: I am not sure whether we should use `floor` here, @kgryte if you can provide your opinion.

Comment (Jordan-Gallivan) [reply lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/native/benchmark.c]: @Pranavchiku since x is type casted to an int32_t, my assumption is it would act as a floor method, so agreed, the floor is redundant

File: lib/node_modules/@stdlib/math/base/special/negalucas/examples/c/example.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/examples/c/example.c:29]:
```suggestion
        printf( "negalucas(%d) = %d\n", i, v );
```
in my opinion.

File: lib/node_modules/@stdlib/math/base/special/negalucas/include.gypi
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/include.gypi]:
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/Makefile]: You need to update copyright years.
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/addon.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/addon.c]: Add new line at end.

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]: Maybe add some more example which returns NaN as well.

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]:
```suggestion
    int32_t an;
    an = (int32_t)stdlib_base_abs( n );
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile
Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile]: Okay, you mixed up two different c benchmarks, create a separate folder `native` inside `c` benchmarks there, and then copy over Makefile.

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/benchmark.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/benchmark.c:90]: Revert this changes.

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile
Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile]: You can take a look at how it works https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/sinh/benchmark/c.

Comment (Jordan-Gallivan) [reply lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile]: that's my bad!  I'll get to fixing it once you all are done with the initial review

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Dropped a few suggestions kindly apply those and do ping me in case of any difficulties :)

**Comment by Jordan-Gallivan** [comment]
@Pranavchiku made the changes per the above messages.  Just to clarify, the copyright date for all files should be 2023 correct?  If not, I can go through and clean those up.

I took out the floor() method call in native/benchmark.c file per the above discussion.  It looks like it worked.  

I added an example in the src/main.c file that returns NaN.  Didn't know if you wanted an extra new line in between the two examples.  If not, that'll be an easy clean up on my end. 

Additionally, I fixed the MakeFiles for the benchmark and benchmark/native.  That was my bust.

All files compiled and all tests passed!  Please let me know if you see anything else that needs fixing.

**Comment by Pranavchiku** [comment]
> Just to clarify, the copyright date for all files should be 2023 correct?

Yes for all those files which you updated and/or added.

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile:4]: This needs to be reverted, as we are not updating this Makefile.

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/Makefile]: Add newline at end of file.

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]: Rather than defining it, try using macro at https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/constants/float64/max-safe-nth-lucas.
You may refer https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/src/main.c.

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
I have left few suggestions kindly have a look at those, along with that `benchmark.native.js` is missing you will need to add that. Thanks for working on this :)

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (Jordan-Gallivan) [reply lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]: @Pranavchiku When i referenced the Max safe nth lucas from the constants, I'm no longer able to get the files to compile.  In the past I've been running the following commands:
make install-node-modules
make examples-c EXAMPLES_FILTER=".*/math/base/special/negalucas/.*"
make benchmark-c BENCHMARKS_FILTER=".*/math/base/special/negalucas/.*"
make install-node-addons NODE_ADDONS_PATTERN=math/base/special/negalucas
make test TESTS_FILTER=".*/math/base/special/negalucas/.*"

is there an intermediate step I need to run after make install-node-modules so the includes can reference constants?

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]: As you are including "stdlib/constants/float64/max_safe_nth_lucas.h" you do not need to explicitly define `STDLIB_CONSTANT_FLOAT64_MAX_SAFE_NTH_LUCAS`
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json]: You will need to add macro to dependency as well.

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]:
```suggestion
        return 0.0 / 0.0; // NaN
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
After adding MACRO to dependency and removing the declaration, I guess you will be able to clear benchmarks and pass tests. LMK if you face any difficulty.

**Comment by Jordan-Gallivan** [comment]
@Pranavchiku just updated the PR with the above changes and all tests passed.  Thanks for your patience, this has been an awesome learning experience.

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/Makefile]:
```suggestion
# Copyright (c) 2018 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/examples/c/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/negalucas/examples/c/Makefile]:
```suggestion
# Copyright (c) 2023 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/tribonacci/benchmark/c/Makefile
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/tribonacci/benchmark/c/Makefile]:
```suggestion
# Copyright (c) 2020 The Stdlib Authors.
```

**Review by Pranavchiku — APPROVED** [review]
LGTM! Thanks @Jordan-Gallivan!! Let @kgryte have a look before merging this.

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/benchmark.c
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/benchmark.c:92]: Thanks for catching this and fixing it!

File: lib/node_modules/@stdlib/math/base/special/negalucas/include/stdlib/math/base/special/negalucas.h
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/include/stdlib/math/base/special/negalucas.h]:
```suggestion
```

We can go ahead and remove this, as this originates from files adhering to older project conventions.

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c:101]:
```suggestion
};

```

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]:
```suggestion
* Computes the nth negaLucas number.
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json]:
```suggestion
        "@stdlib/math/base/special/labs",
```

We have a 32-bit variant for use in C.

File: lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json]:
```suggestion
        "@stdlib/math/base/special/labs",
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/manifest.json]:
```suggestion
        "@stdlib/math/base/special/labs",
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]:
```suggestion
#include "stdlib/math/base/special/labs.h"
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/src/main.c]:
```suggestion
    int32_t an = stdlib_base_labs( n );
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js]:
```suggestion
	for ( i = -77; i > -500; i-- ) {
```

As is, this loop will never iterate.

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js]:
```suggestion
tape( 'if provided nonpositive integers less than `-76`, the function returns `NaN`', opts, function test( t ) {
```

We should maintain parity with `test.js`.

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:74]:
```suggestion
tape( 'the function returns the nth negaLucas number', opts, function test( t ) {
```

This test block is not testing the Lucas relationship.

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js]:
```suggestion
		t.strictEqual( v, NEGALUCAS[ abs( i ) ], 'returns the '+i+'th negaLucas number' );
```

This test is broken, as `i` is nonpositive, but the indices for `NEGALUCAS` must be nonnegative.

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:29]:
```suggestion
var pow = require( '@stdlib/math/base/special/pow' );
var abs = require( '@stdlib/math/base/special/abs' );
var tryRequire = require( '@stdlib/utils/try-require' );
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:85]:
```suggestion

tape( 'the function returns the nth negaLucas number (Lucas relationship)', opts, function test( t ) {
	var v;
	var f;
	var i;
	for ( i = 0; i > -77; i-- ) {
		v = negalucas( i );
		f = pow( -1.0, abs(i) ) * lucas( abs(i) );
		t.strictEqual( v, f, 'returns the '+i+'th negaLucas number' );
	}
	t.end();
});

```

Add missing tests for the Lucas relationship. In short, our `test.native.js` files should more or less match our `test.js` files. Meaning, calling into C from JavaScript should yield the same results as if we had just invoked a JavaScript implementation. As such, tests

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js]:
```suggestion
```

This is one where we don't expect parity between JS and C. C expects 32-bit ints, so we test accordingly.

File: lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/test/test.native.js:27]:
```suggestion
var abs = require( '@stdlib/math/base/special/abs' );
var lucas = require( '@stdlib/math/base/special/lucas' );
```

**Review by kgryte — APPROVED** [review]
LGTM. Will merge once CI passes. Thanks, @Jordan-Gallivan and @Pranavchiku!

File: lib/node_modules/@stdlib/math/base/special/negalucas/include/stdlib/math/base/special/negalucas.h
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/include/stdlib/math/base/special/negalucas.h]:
```suggestion
* Computes the nth negaLucas number.
```

File: lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/native/benchmark.c
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/negalucas/benchmark/c/native/benchmark.c]: This was addressed.

**Review by kgryte — CHANGES_REQUESTED** [review]
@Jordan-Gallivan Looking again. Looks like we're missing updates to the README. Would you mind documenting the C APIs in the `negalucas` README file? See Fibonacci for an example.

**Comment by Jordan-Gallivan** [comment]
@kgryte I'd be happy to.  Thanks for cleaning up some of the slop in my code!

I noticed you changed from abs to labs.  Would you mind if I ask why (continuing the learning on my end)?

**Comment by kgryte** [comment]
Happy to explain.

1. **Avoiding unnecessary type conversions.** The signature for `stdlib_base_abs` is

    ```c
    double stdlib_base_abs( double x );
    ```

    So `double` in and `double` out. The input argument for `stdlib_base_negalucas` is `int32_t`. Accordingly, when passing the input argument to `stdlib_base_abs`, we need to perform an implicit type conversion `(double)<int32_t>`. And then we were performing an explicit type conversion of the return value

    ```c
    int32_t an = (int32_t)stdlib_base_abs(...)
    ```

    In which case, we're performing two type conversions, which is less than optimal.

    In contrast, the signature for `stdlib_base_labs` is

    ```c
    int32_t stdlib_base_labs( int32_t x );
    ```

    which matches our desired input and output dtypes and thus avoids the undesired type conversions.

2. **Avoiding unnecessary branching.** Because of floating-point, `stdlib_base_abs` needs to perform additional branching logic, particularly with regard to positive and negative zero. These concerns are not applicable when working with integer dtypes. Instead, we can perform bit manipulation, which is what `stdlib_base_labs` does and which **can** be easier for a compiler to optimize.

Hope this helps!

**Comment by Jordan-Gallivan** [comment]
@kgryte Thank you so much for the explnation.  I'm really digging what you're doing with this project and it's taught me a a lot so far in the little that I've been able to contribute.  I just updated the README for this and I'll pop back to the tribonacci to get that knocked out as well.

**Comment by kgryte** [comment]
@Jordan-Gallivan Glad to help. I've already updated [tribonacci](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/@stdlib/math/base/special/tribonacci#c-apis), so no need to worry about that. Feel free to move another open issue.

File: lib/node_modules/@stdlib/math/base/special/negalucas/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/negalucas/README.md]:
```suggestion
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @Jordan-Gallivan!

**Pull request merged**
